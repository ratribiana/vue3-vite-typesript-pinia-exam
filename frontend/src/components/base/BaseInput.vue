<script setup lang="ts">
import { useField } from 'vee-validate';

const props = defineProps({
  name: String,
  type: String,
  id: String,
  placeholder: String,
  disabled: Boolean,
  label: String,
  customClass: {
    type: String,
    default: '',
    description: 'Additional custom class to append to the text field',
  },
});

// The `name` is returned in a function because we want to make sure it stays reactive
// If the name changes you want `useField` to be able to pick it up
const { value, errorMessage } = useField(() => props.name);
</script>
<template>
  <label v-if="label" :for="name" class="block mb-2 text-sm font-medium text-gray-900 dark:text-white"
    >{{ label }} <small v-if="$attrs.required == 'required'" class="required">*</small></label
  >
  <input
    v-model="value"
    :name="name"
    :type="type || 'text'"
    :id="id"
    :placeholder="placeholder"
    :disabled="disabled"
    :class="[{ error: errorMessage }, customClass]"
    v-bind="$attrs"
    class="normal-case"
  />
  <small class="text-red dark:text-[#f78181]">{{ errorMessage }}</small>
</template>
