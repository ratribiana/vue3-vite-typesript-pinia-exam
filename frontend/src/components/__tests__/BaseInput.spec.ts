import { describe, expect, it } from 'vitest';
import { mount } from '@vue/test-utils';
import BaseInput from '@/components/base/BaseInput';

describe('BaseInput', () => {
  it('renders correctly with name, type, id, label and placeholder props', () => {
    const wrapper = mount(BaseInput, {
      props: {
        name: 'customInput',
        type: 'password',
        id: 'customId',
        placeholder: 'Enter your password',
        label: 'Password',
      }
    });

    // Assert on the label existence
    const label = wrapper.find('label');
    expect(label.exists()).toBe(true);
    expect(label.text()).toContain('Password');

    // Assert on the input existence and attributes
    const input = wrapper.find('input');
    expect(input.exists()).toBe(true);
    expect(input.attributes('name')).toBe('customInput');
    expect(input.attributes('type')).toBe('password');
    expect(input.attributes('id')).toBe('customId');
    expect(input.attributes('placeholder')).toBe('Enter your password');
  });

  it('renders correctly with all props', () => {
    const wrapper = mount(BaseInput, {
      props: {
        name: 'customInput',
        type: 'password',
        id: 'customId',
        placeholder: 'Enter your password',
        disabled: false,
        label: 'Password',
        customClass: 'custom-input',
      },
      attrs: {
        required: 'required',
      },
    });

    // Assert on the rendered HTML
    expect(wrapper.html()).toMatchSnapshot();

    // Assert on the label existence
    const label = wrapper.find('label');
    expect(label.exists()).toBe(true);
    expect(label.text()).toContain('Password');

    // Assert on the small element existence and content
    const asterisk = wrapper.find('.required');
    expect(asterisk.exists()).toBe(true);
    expect(asterisk.text()).toBe('*');

    // Assert on the input existence and attributes
    const input = wrapper.find('input');

    expect(input.exists()).toBe(true);
    expect(input.attributes('name')).toBe('customInput');
    expect(input.attributes('type')).toBe('password');
    expect(input.attributes('id')).toBe('customId');
    expect(input.attributes('placeholder')).toBe('Enter your password');
    expect(input.attributes('disabled')).toBe(undefined); // Check that the 'disabled' attribute is not present
    expect(input.attributes('required')).toBe(''); // Check that the 'required' attribute is present
    expect(input.classes()).toContain('custom-input');
    
    // Assert on the error message existence
    const errorMessage = wrapper.find('.text-red');
    expect(errorMessage.exists()).toBe(true);
  });
});
