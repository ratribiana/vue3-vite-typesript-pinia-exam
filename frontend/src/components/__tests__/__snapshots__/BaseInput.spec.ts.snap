// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BaseInput > renders correctly with all props 1`] = `
"<label for="customInput" class="block mb-2 text-sm font-medium text-gray-900 dark:text-white">Password <small class="required">*</small></label>
<input name="customInput" type="password" id="customId" placeholder="Enter your password" class="custom-input normal-case" required="">
<small class="text-red dark:text-[#f78181]"></small>"
`;
